### ISSUE TITLE
doc for API loadbalancer 

### ISSUE THREAD

**Issue by jichenjc** [kubernetes-sigs/cluster-api-provider-openstack#422]
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]
we merged API loadbalancer code and we need additional doc 

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Comment by jichenjc** [comment]
@sbueringer could you please provide some help on your local yaml settings?

I tried my local env a little bit and If a sample configuration existing will be much helpful to me on the APIloadbalancer stuff...thanks

**Comment by jichenjc** [comment]
I can see the LB/listener etc created but I got error like below...

E0805 02:53:26.896054       1 actuator.go:427] Machine error openstack-node-t97sp: error creating Openstack instance: unable to create manager for restConfig: Get https://9.20.206.79:6443/api?timeout=32s: dial tcp 9.20.206.79:6443: i/o timeout
W0805 02:53:26.896313       1 machine_controller.go:231] Failed to create machine "openstack-node-t97sp": error creating Openstack instance: unable to create manager for restConfig: Get https://9.20.206.79:6443/api?timeout=32s: dial tcp 9.20.206.79:6443: i/o timeout

**Comment by jichenjc** [comment]
I can see the LB/listener etc created but I got error like below...

E0805 02:53:26.896054       1 actuator.go:427] Machine error openstack-node-t97sp: error creating Openstack instance: unable to create manager for restConfig: Get https://9.20.206.79:6443/api?timeout=32s: dial tcp 9.20.206.79:6443: i/o timeout
W0805 02:53:26.896313       1 machine_controller.go:231] Failed to create machine "openstack-node-t97sp": error creating Openstack instance: unable to create manager for restConfig: Get https://9.20.206.79:6443/api?timeout=32s: dial tcp 9.20.206.79:6443: i/o timeout

**Comment by sbueringer** [comment]
Do you know if the APIServer is up and the IP of your LB is generally reachable from inside your cluster?
(We have the problem that in some of our environments FloatingIPs cannot be accessed from inside the cluster)

My configuration is:
````
      managedAPIServerLoadBalancer: true
      apiServerLoadBalancerFloatingIP: 75.12.112.86
      apiServerLoadBalancerPort: 6443
      apiServerLoadBalancerAdditionalPorts:
        - 22
````

**Comment by jichenjc** [comment]
I had same settings to you and I used the same master ip I used previously which I can use it as master ip to connect to..

**Comment by sbueringer** [comment]
And you did not also use it as FloatingIP on the master vm right?

If the APIServer itself is up than I think it must be that the IP is not routable inside the cluster. How did the non-lb case connect to the APIServer? Probably also via the FloatingIP, right? That would mean that's also not the problem

**Comment by jichenjc** [comment]
at first I assigned but I realized I can't do that, so I didn't assign floating ip on the master and test again

if no APIsever, the floating ip will be assigned to master and everything is fine. I knew octavia as well, as it will create a machine and only visiable to admin and haproxy will run inside 
the thing I don't know is LBaaSv2 using neutron , does it create a similar VM and put haproxy or using other magic to host the haproxy or its similar work?( I will dig more here as well..but if you can provide some some guide it will be helpful to me )

**Comment by hidekazuna** [comment]
It seems that clusterProviderSpec.ExternalNetworkID is also required.

https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/master/pkg/cloud/openstack/services/networking/loadbalancer.go#L23-L25

**Comment by sbueringer** [comment]
> if no APIsever, the floating ip will be assigned to master and everything is fine. I knew octavia as well, as it will create a machine and only visiable to admin and haproxy will run inside
> the thing I don't know is LBaaSv2 using neutron , does it create a similar VM and put haproxy or using other magic to host the haproxy or its similar work?( I will dig more here as well..but if you can provide some some guide it will be helpful to me )

Our Neutron LBaasV2 implements this stuff via NSX-V and therefore also uses HAProxy under the hood. I'm not sure, but I think OpenStack itself doesn't specify how this must be implemented and it should depend on the distribution.

@hidekazuna Yes you're right externalNetworkID is also required, but because it's also required to create the Router I though it's set :)

My complete (anonymized) configuration:
````
apiVersion: "cluster.k8s.io/v1alpha1"
kind: Cluster
metadata:
  name: test
spec:
  clusterNetwork:
    services:
      cidrBlocks: ["10.254.0.0/16"]
    pods:
      cidrBlocks: ["10.6.0.0/16"]
    serviceDomain: "cluster.local"
  providerSpec:
    value:
      apiVersion: "openstackproviderconfig/v1alpha1"
      kind: "OpenstackProviderSpec"
      nodeCidr: "10.6.0.0/24"
      managedAPIServerLoadBalancer: true
      apiServerLoadBalancerFloatingIP: 75.12.112.86
      apiServerLoadBalancerPort: 6443
      apiServerLoadBalancerAdditionalPorts:
        - 22
      dnsNameservers:
        - 75.12.7.144
      externalNetworkId: asd12as-c542-410f-929c-ec9b1830f40c
      externalRouterIPs:
        - fixedIP: 75.12.112.81
          subnet:
            filter:
              name: test
      managedSecurityGroups: false
      disableServerTags: true
````

Can you post the output of the following commands?

````
neutron lbaas-loadbalancer-list
+--------------------------------------+--------------------------------------+-------------+---------------------+------------+
| id                                   | name                                 | vip_address | provisioning_status | provider   |
+--------------------------------------+--------------------------------------+-------------+---------------------+------------+
| bc1bbac9-9514-4d89-bc55-3f73a25e6037 | k8s-cluster-test-os1pi020-kubeapi | 10.6.0.12   | ACTIVE              | vmwareedge |
+--------------------------------------+--------------------------------------+-------------+---------------------+------------+
openstack floating ip list
+--------------------------------------+---------------------+------------------+--------------------------------------+--------------------------------------+----------------------------------+
| ID                                   | Floating IP Address | Fixed IP Address | Port                                 | Floating Network                     | Project                          |
+--------------------------------------+---------------------+------------------+--------------------------------------+--------------------------------------+----------------------------------+
| 32a16bed-52f0-427b-ac7c-81c1a3315423 | 75.12.112.86        | 10.6.0.12        | b7e3a786-0daa-4bef-88a6-5e0065363e2f | asd12as-c542-410f-929c-ec9b1830f40c | b9abf84da3944b85a61cab7aecf97d77 |
+--------------------------------------+---------------------+------------------+--------------------------------------+--------------------------------------+----------------------------------+
neutron lbaas-pool-list
+--------------------------------------+-------------------------------------------+--------------+----------+----------------+
| id                                   | name                                      | lb_algorithm | protocol | admin_state_up |
+--------------------------------------+-------------------------------------------+--------------+----------+----------------+
| f0e0ee1e-6392-4826-8e23-88c46527057f | k8s-cluster-default-test-kubeapi-22   | ROUND_ROBIN  | TCP      | True           |
| fba4cfca-9647-4fe8-8a09-30f190bad4f4 | k8s-cluster-default-test-kubeapi-6443 | ROUND_ROBIN  | TCP      | True           |
+--------------------------------------+-------------------------------------------+--------------+----------+----------------+
neutron lbaas-member-list k8s-cluster-default-test-kubeapi-6443

+--------------------------------------+-------------------------------------------------------------------+----------+---------------+--------+--------------------------------------+----------------+
| id                                   | name                                                              | address  | protocol_port | weight | subnet_id                            | admin_state_up |
+--------------------------------------+-------------------------------------------------------------------+----------+---------------+--------+--------------------------------------+----------------+
| 5ed716ae-6cd7-408c-95d4-84f2ce39da26 | k8s-cluster-default-test-kubeapi-6443-os1pi020-kube-master-01 | 10.6.0.4 |          6443 |      1 | 732a4431-9063-46e7-b753-cf6b4d4bfea4 | True           |
| 386704b4-faee-4ecb-a8ae-0f467a104596 | k8s-cluster-default-test-kubeapi-6443-os1pi020-kube-master-02 | 10.6.0.7 |          6443 |      1 | 732a4431-9063-46e7-b753-cf6b4d4bfea4 | True           |
+--------------------------------------+-------------------------------------------------------------------+----------+---------------+--------+--------------------------------------+----------------+
````

**Comment by jichenjc** [comment]
thanks a lot for your detail info , I believe the 'NSX-V' makes the magic and I don't know the detail implementation of my env's LBaaSv2 ... as all the CLI output above I have simliar one in my env but I guess it's only inside API db and nothing real implemented..anyway, I need find an octavia env :)

**Comment by jichenjc** [comment]
and , we may need some doc update in order to close this issue :)

**Comment by sbueringer** [comment]
> and , we may need some doc update in order to close this issue :)

Yup, that's fine. Just wanted to find out what's wrong :P. Okay so documentation would be like a section about `How to configure APIServer LoadBalancer`? And more or less mentioning the properties from above and that the LB needs a separate FloatingIP, right?

**Comment by jichenjc** [comment]
yes, exactly, I think we need provide new comer something to refer to instead of asking simliar stuffs in the issue or slack channel, and the doc don't need to be perfect, just something for people to start with then we can keep enhancing it in the future

**Comment by sbueringer** [comment]
Alright. I'll open a PR soon for it.

**Comment by jichenjc** [comment]
@hidekazuna @sbueringer I am not sure whether good place to discuss here,anyway post my findings:

The LB issue I have is actually comes from NodeCIDR, I set NodeCIDR to 172.17.31.1/24 while the networks I am using for boot vm is 10.43.0.1/24 (precreated in my environment)
so I manually create a LB and when create the LB, I select the subnet of 10.43.0.1/24 and the floating ip of the LB is '9.x.x.x' then from internal and out the cluster I can both wget 9.x.x.x (the traffic did forward to the 10.43.x.x vm)

so I tried following but not finished yet, the idea and question listed below:
```
--- a/pkg/cloud/openstack/cluster/actuator.go
+++ b/pkg/cloud/openstack/cluster/actuator.go
@@ -93,7 +93,12 @@ func (a *Actuator) Reconcile(cluster *clusterv1.Cluster) error {

        klog.Infof("Reconciling network components for cluster %s", clusterName)
        if clusterProviderSpec.NodeCIDR == "" {
-               klog.V(4).Infof("No need to reconcile network for cluster %s", clusterName)
+               if clusterProviderSpec.ManagedAPIServerLoadBalancer {
+                        err = networkingService.ReconcileLoadBalancer(clusterName, clusterProviderSpec, clusterProviderStatus)
+                        if err != nil {
+                                return errors.Errorf("failed to reconcile load balancer: %v", err)
+                        }
+                }
```

```
--- a/pkg/cloud/openstack/services/networking/loadbalancer.go
+++ b/pkg/cloud/openstack/services/networking/loadbalancer.go
@@ -15,6 +15,7 @@ import (
        constants "sigs.k8s.io/cluster-api-provider-openstack/pkg/cloud/openstack/contants"
        clusterv1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
        "sigs.k8s.io/cluster-api/pkg/util"
+       openstackconfigv1 "sigs.k8s.io/cluster-api-provider-openstack/pkg/apis/openstackproviderconfig/v1alpha1"
        "time"
 )

@@ -45,7 +46,8 @@ func (s *Service) ReconcileLoadBalancer(clusterName string, clusterProviderSpec
                klog.Infof("Creating loadbalancer %s", loadBalancerName)
                lbCreateOpts := loadbalancers.CreateOpts{
                        Name:        loadBalancerName,
-                       VipSubnetID: clusterProviderStatus.Network.Subnet.ID,
+                       //VipSubnetID: clusterProviderStatus.Network.Subnet.ID,
+                       VipSubnetID: "bef4301e-5842-4136-9254-e32396077e59",
                }
```

1) without a NodeCIDR, we can't create a LB ? then how about an existing network/subnet case? why should we create network in any case? can we allow a predefined network ? 
2) according to 1 above, we made the LB in Network, should this be outside of the network definition as we might have LB but don't have cluster network defined?
```
type OpenstackClusterProviderStatus struct {
        metav1.TypeMeta   `json:",inline"`
        metav1.ObjectMeta `json:"metadata,omitempty"`

        // Network contains all information about the created OpenStack Network.
        // It includes Subnets and Router.
        Network *Network `json:"network,omitempty"`


// Network represents basic information about the associated OpenStach Neutron Network
type Network struct {
        Name string `json:"name"`
        ID   string `json:"id"`

        Subnet *Subnet `json:"subnet,omitempty"`
        Router *Router `json:"router,omitempty"`

        // Be careful when using APIServerLoadBalancer, because this field is optional and therefore not
        // set in all cases
        APIServerLoadBalancer *LoadBalancer `json:"apiServerLoadBalancer,omitempty"`
}

```


long story short, I believe we should support a case that we have predefined network (subnet) for VM provisioning case ....

**Comment by hidekazuna** [comment]
@jichenjc @sbueringer 

> 1. without a NodeCIDR, we can't create a LB ? then how about an existing network/subnet case? why should we create network in any case? can we allow a predefined network ?

Sure. without a NodeCIDR, we can't create a LB. I tried to move networkingService.ReconcileLoadBalancer outside of if clusterProviderSpec.NodeCIDR == "" statement, but result in error due to clusterProviderStatus.Network.Subnet.ID missing. Now I understand it is by design.

> 2. according to 1 above, we made the LB in Network, should this be outside of the network definition as we might have LB but don't have cluster network defined?

Actually cloud-provider-openstack has separate definition. but I am not sure we also have to do so.

> long story short, I believe we should support a case that we have predefined network (subnet) for VM provisioning case ....

+1 in the future. Just now or after finishing #382 ? I think now we are under #382 so that we finish by successful scenario at first then follow up other scenarios like predefined case. @sbueringer What do you think?

**Comment by jichenjc** [comment]
I am ok on both after or before, just want to make sure it's agreed by folks...

so the questions still remaining: ` without a NodeCIDR, we can't create a LB?` this will impact the design and coding etc..

**Comment by sbueringer** [comment]
First of all I think sooner or later (maybe after the migration to v1alpha2) we should write down what we require from an OpenStack and which kind of scenarios we want to support (e.g. external Network already created, all Networks/Subnets already preprovisioned, FloatingIPs already there,...) Then we should re-design our configuration.

Regarding the current problem:
@jichenjc Your case is that all networks are already created before right? That means if you could specify or match on your existing networks it should all be fine.

What we're currently doing:
* cluster actuator:
  * if NodeCidr is specified > network/subnet/router will be created and the results stored inside the status
* machine actuator:
  * It's possible to specify "matchers" in the Machinespec to find the networks which have been created by the cluster actuator

I think this problem can be solved like this:
* cluster actuator:
  * create the network/subnet/router if necessary if not reuse the existing ones
* machine actuator:
  * just use the network/subnet from the cluster status (do we have the use-case to use another network in every machine? I don't think so.

Than the only question is how to configure that on the Cluster level. The easiest solution for now would be to just move the networks configuration from machine to cluster. This means that on a cluster level it's possible to either:
* specify a nodeCidr and network/subnet/router will be created and added to the status
* or specify the following and the network/subnet will be found and added to the status:
````
networks:
          - filter:
              name: <network-name>
            subnets:
              - filter:
                  name: <subnet-name>
````

Would this enable your use case?

**Comment by hidekazuna** [comment]
I found one thine to note. OpenStack user have to admin priviledge to create floating ip specified by apiServerLoadBalancerFloatingIP in cluster.yaml. 

Without admin priviledge, I got the following error.

```
E0807 04:45:50.050676       1 cluster_controller.go:146] Error reconciling cluster object test401; failed to reconcile load balancer: error allocating floating IP: Request forbidden: [POST http://controller.hidekazuna.test:9696/v2.0/floatingips], error message: {"NeutronError": {"type": "PolicyNotAuthorized", "message": "(rule:create_floatingip and rule:create_floatingip:floating_ip_address) is disallowed by policy", "detail": ""}}
```

**Comment by jichenjc** [comment]
this should be configurable... at least in my env I am not admin but I can create lb 
it should be defined in neutron policy or octavia policy..

**Comment by sbueringer** [comment]
In my opinion it's already implemented correctly. It only tries to create the FloatingIP if the FloatingIP isn't already there: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/master/pkg/cloud/openstack/services/networking/loadbalancer.go#L327

So if you make sure the FloatingIP exists, it should just use that (similar to the way machines are associated to existing FloatingIPs)

**Comment by jichenjc** [comment]
> I think we should also add the create-if-it-doesn't-alread-exist logic to the machine FloatingIPs (currently it just fails if the FloatingIP is not created before)

The problem is seems we need the floating ip in order to create a kubeconfig file and communicate with the floating ip? I will open an issue and track it , this seems need thorough thought first

**Comment by hidekazuna** [comment]
OK, floating ip issue moves to #427

**Comment by sbueringer** [comment]
> > I think we should also add the create-if-it-doesn't-alread-exist logic to the machine FloatingIPs (currently it just fails if the FloatingIP is not created before)
> 
> The problem is seems we need the floating ip in order to create a kubeconfig file and communicate with the floating ip? I will open an issue and track it , this seems need thorough thought first

Yup. Afaik having FloatingIPs is always a requirement. The best we can and should do is to make it possible to use them if the already exist and just create them when necessary.

**Comment by sbueringer** [comment]
@jichenjc I would open a PR the next view days to move the networks setting to the cluster resource to cleanup the configuration mess :)

**Comment by jichenjc** [comment]
ok, you mean , the networks from the machine configuration to cluster configuration? I think it's a good idea but we might need some thorough thought here, whether we have any use case that machines need to be different net/subnet (e.g HA, affinity etc), I can't image one but I don't know whether there might be..

**Comment by sbueringer** [comment]
Yes. We can also just add it to the Cluster CRD for now and use it to write the cluster status. We can also use this as fallback if networks on the Machine CRD is not specified.
